For finding cluster dup

jaysim opened the pull request:
- fix index entry length corruption

- check cluster duplication, and remove possibility of cluster duplication in fsck
- remove unnecessary function.
- change lost+found added filename renaming policy

jaysim committed:
ntfsck: add fsck.c/fsck.h in libntfs-3g

1st step for checking bitmap duplication.
separate some functions related with fsck bitmap from ntfsck.c
add new file fsck.c/fsck.h in library directory

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: rename function/clean macro definition of bitmap allocation

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: add set/clear fsck lcn bitmap in lcnalloc.c

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: remove uncessary update_lcn_bitmap() call

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: add check bitmap duplication

Add lcn bitmap set when checking non-resident attribute.
Remove unnecessary paramter of ntfsck_check_runlist() and unnecessary
function.

Skip unnecessary calling function ntfsck_check_non_resident_attr() in
ntfsck_check_inode_non_resident(). Because all of same attributes of inode
are already checked in ntfsck_check_inode_non_resident(). So, there is
no need to call ntfsck_check_inode_non_resident() for same attribute.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: fix returning data of corrupted runlist on fsck

ntfs_mapping_pairs_decompress_on_fsck() function try to return
maximal valid runlist data. But, when there are corrupted runlists data
which have length is not zero, but offset is zero.
ntfs_mapping_pairs_decompress_on_fsck() return duplicated runlist data.

Because ntfs_mapping_pairs_decompress_i() function does not consider above
case(length is not zero, offset is zero), So it return runlist data
which has duplicated vcn/lcn value.

Before fixing it, we should consider about $Boot/$Data attribute.
The first entry offset of $Boot/$Data runlist is always zero,
because bootsector located at cluster 0.
fsck should need to distinguish it.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: optimize bitmap duplication check logic

Use test_and_set() function to check cluster duplication.

TODO: need more optimization

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
libntfs: add function to find zero/set bit

Add function to find first set/zero bit in word.
Add function to find next set/zero bit in word.
Almost code from linux find.h but ffsl().
Function ffsl() in library returns 1 ~ 32 or 64, but kernel ffs() return
0 to 31, so function has been modified to match return value like
kernel

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
libntfs: add function to get mftbmp block of fsck

The bitmap block of fsck pointer is in ntfs_volume structure.
Add function to get mft record bitmap block using mft number.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: modify checking index bitmap

The index bitmap of directory inode is localted in ntfs_inode structure.
Add index bitmap size field in ntfs_inode, that is used in index bitmap
checking routine.

Fix bug in ntfsck_set_index_bitmap() function.
(local variable vcn is not initialized)

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: Remove bitmap duplication possibility of ntfsck

There is a possibility that ntfsck may cause bitmap duplication.
This patch remove it.
- move checking lost found directory after applying bitmap.
- separate checking bitmap and repairing orphaned mft records.
- move repairing orphaned mft records to after bitmap checking.
- before checking orphaned mft records, repaired mft/cluster bitmap is
  applied to disk.

Remove unnecessary functions.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: check backup boot sector check and set bitmap

NTFS backup boot sector located in last cluster/sector, or in the middle
of cluster. When it located in the middle, it's cluster bitmap should be
set not to allocate by others.
After it checked, set cluster bitmap, because ntfs set bitmap of last or middle
of cluster which is located backup boot sector.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
libntfs-3g: fix some bugs and improve repairing

When checking index block, other checking is good, then will modify
magic number in case magic number is wrong.

When open system mft record, will set it used if it set unused.
Because essential system mft should never be deleted.

Fix memory leak.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: modify and debug orphaned inode check logic

Modify ntfsck_find_and_check_index() logic. If found index in parent
inode, then check mft and sequence number before exit or checking. (added)

When check inode in ntfsck_find_and_check_index(), if the inode had been
checked already, just check file type and $FILE_NAME attr. do not call
ntfsck_check_inode()
(to prevent with checking cluster duplication false positive)

Just exit if base mft bitmap is not set on disk in
ntfsck_verify_mft_record(). And if open failed inode is extent inode,
just return, do not add in orphaned inode candidate list.
Before add orphaned candidate list,

Add helper function ntfsck_open_inode/ntfsck_close_inode() not to open
again already opened inode on volume mount.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: initialize index structure when check failed

If found some errors in index block, then just initialize index root/
index block / index bitmap structure.
Those files which is not tested will be checked in orphan mft routine.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

namjaejeon commented on libntfs-3g/fsck.c:
```diff
@@ -0,0 +1,155 @@
+#ifdef HAVE_CONFIG_H
+#include "config.h"
```
Is there any reason why fsck codes move to library ?

jaysim replied:
because library functions should call function in fsck.c, like as ntfs_cluster_alloc() / ntfs_cluster_free().
those functions set bitmap on disk.
if those were called in fsck context, they should set fsck bitmap also to synchronize bitmap between on disk bitmap and fsck in memory bitmap.

jaysim committed:
ntfsck: fix vcn calculation in case cluster size is less than 4K

Index allocation block size is always 4K. but cluster size can be
variable, also vcn size do.

when cluster size is smaller than 4K, vcn size is same as cluster size.
But cluster size is larget than 4K, vcn size set to 512 byte.

Fix wrong calculated vcn valud in fsck.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim committed:
ntfsck: clear fsck lcnbmp when index block is initialized

When index block is initialized, cluster allocated as $INDEX_ALLOCATION
should be cleared in fsck lcnbmp.

Signed-off-by: JaeHoon Sim <jay.sim@lge.com>

jaysim commented:
#23 will include these patchsets. close this PR
